drivers/firmware/ti_sci/ti_sci.c

```diff
@@ -0,0 +1,1660 @@
+	int ret;
+
+	if (!req_state && !curr_state) {
+		return -EINVAL;
+	}
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	ret = ti_sci_cmd_get_clock_state(dev, dev_id, clk_id, &r_state, &c_state);
+	if (ret) {
+		return ret;
+	}
+
+	if (req_state) {
+		*req_state = (r_state == MSG_CLOCK_SW_STATE_UNREQ);
+	}
+	if (curr_state) {
+		*curr_state = (c_state == MSG_CLOCK_HW_STATE_NOT_READY);
+	}
+	return 0;
+}
+
+int ti_sci_cmd_clk_get_match_freq(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+				  uint64_t min_freq, uint64_t target_freq, uint64_t max_freq,
+				  uint64_t *match_freq)
+{
+	struct ti_sci_msg_resp_query_clock_freq resp;
+	struct ti_sci_msg_req_query_clock_freq req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_QUERY_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), &resp, sizeof(resp));
+
+	if (!xfer) {
+		LOG_ERR("Failed to setup transfer");
+		return -EINVAL;
+	}
+
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+	req.min_freq_hz = min_freq;
+	req.target_freq_hz = target_freq;
+	req.max_freq_hz = max_freq;
+
+	ret = ti_sci_do_xfer(dev, xfer);
+	if (ret) {
+		return ret;
+	}
+
+	*match_freq = resp.freq_hz;
+
+	return 0;
+}
+
+int ti_sci_cmd_clk_set_freq(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			    uint64_t min_freq, uint64_t target_freq, uint64_t max_freq)
+{
+	struct ti_sci_msg_req_set_clock_freq req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	if (!dev) {
+		return -EINVAL;
+	}
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), NULL, 0);
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+	req.min_freq_hz = min_freq;
+	req.target_freq_hz = target_freq;
+	req.max_freq_hz = max_freq;
+
+	ret = ti_sci_do_xfer(dev, xfer);
+
+	return ret;
+}
+
+int ti_sci_cmd_clk_get_freq(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			    uint64_t *freq)
+{
+	struct ti_sci_msg_resp_get_clock_freq resp;
+	struct ti_sci_msg_req_get_clock_freq req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_GET_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), &resp, sizeof(resp));
+
+	if (!xfer) {
+		LOG_ERR("Failed to setup transfer");
+		return -EINVAL;
+	}
+
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+
+	ret = ti_sci_do_xfer(dev, xfer);
+	if (ret) {
+		LOG_ERR("Failed to get clock frequency (ret=%d)", ret);
+		return ret;
+	}
+
+	*freq = resp.freq_hz;
+
+	return 0;
+}
+
+int ti_sci_set_clock_state(const struct device *dev, uint32_t dev_id, uint8_t clk_id,
+			   uint32_t flags, uint8_t state)
+{
+	struct ti_sci_msg_req_set_clock_state req;
+	struct ti_sci_xfer *xfer;
+	int ret = 0;
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_CLOCK_STATE,
+				     flags | TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), NULL, 0);
```

Comment:
You should still have a response (not `NULL, 0`), even if it is just a `struct ti_sci_msg_hdr hdr`, that way you do all the normal response checks in `ti_sci_do_xfer()`, like sequence.